File: frontend/src/lib/chat/scheduled-events.ts
import { api } from '$lib/api/client';
import { completeScannedMediaResource } from '$lib/media/scanned';
import { uploadObject, type UploadTicket } from '$lib/media/uploads';
import type { GuildMemberSummary, UserSummary } from './types';

export const ScheduledEventStatus = {
  scheduled: 1,
  active: 2,
  completed: 3,
  canceled: 4
} as const;

export const ScheduledEventEntityType = {
  stage: 1,
  voice: 2,
  external: 3
} as const;

export type ScheduledEventStatusValue =
  (typeof ScheduledEventStatus)[keyof typeof ScheduledEventStatus];
export type ScheduledEventEntityTypeValue =
  (typeof ScheduledEventEntityType)[keyof typeof ScheduledEventEntityType];

export const ScheduledEventRecurrencePreset = {
  none: 'none',
  daily: 'daily',
  weekly: 'weekly',
  biweekly: 'biweekly',
  monthly: 'monthly',
  yearly: 'yearly'
} as const;

export const ScheduledEventRecurrenceFrequency = {
  yearly: 0,
  monthly: 1,
  weekly: 2,
  daily: 3
} as const;

export type ScheduledEventRecurrencePresetValue =
  (typeof ScheduledEventRecurrencePreset)[keyof typeof ScheduledEventRecurrencePreset];

export interface ScheduledEventRecurrenceRule {
  start: string;
  end: null;
  frequency: 0 | 1 | 2 | 3;
  interval: 1 | 2;
  by_weekday?: number[] | null;
  by_n_weekday?: { n: number; day: number }[] | null;
  by_month?: number[] | null;
  by_month_day?: number[] | null;
  by_year_day?: null;
  count?: null;
}

export interface ScheduledEvent {
  id: string;
  origin_domain: string;
  guild_id: string;
  guild_domain: string;
  channel_id: string | null;
  channel_domain: string | null;
  creator_id: string;
  creator_domain: string;
  creator?: UserSummary;
  name: string;
  description: string | null;
  scheduled_start_time: string;
  scheduled_end_time: string | null;
  privacy_level: 2;
  status: ScheduledEventStatusValue;
  entity_type: ScheduledEventEntityTypeValue;
  entity_id: string | null;
  entity_domain: string | null;
  entity_metadata: { location: string } | null;
  recurrence_rule: ScheduledEventRecurrenceRule | null;
  image: string | null;
  created_at: string;
  updated_at: string;
  version: string;
  user_count?: number;
  me_subscribed?: boolean;
}

export interface ScheduledEventUser {
  guild_scheduled_event_id: string;
  guild_scheduled_event_domain: string;
  user: UserSummary;
  member: GuildMemberSummary | null;
  subscribed_at: string;
}

export interface ScheduledEventDraft {
  name: string;
  description: string;
  entityType: ScheduledEventEntityTypeValue;
  channelRef: string;
  location: string;
  startTime: string;
  endTime: string;
  recurrence?: ScheduledEventRecurrencePresetValue;
}

export function scheduledEventRef(event: Pick<ScheduledEvent, 'id' | 'origin_domain'>): string {
  return `${event.id}@${event.origin_domain}`;
}

export function scheduledEventSubscriptionState(
  event: Pick<ScheduledEvent, 'me_subscribed'>,
  localOverride?: boolean
): boolean {
  return localOverride ?? event.me_subscribed ?? false;
}

export function eventChannelRef(
  event: Pick<ScheduledEvent, 'channel_id' | 'channel_domain'>
): string | null {
  return event.channel_id && event.channel_domain
    ? `${event.channel_id}@${event.channel_domain}`
    : null;
}

export function scheduledEventStatusLabel(status: ScheduledEventStatusValue): string {
  if (status === ScheduledEventStatus.active) return 'Live';
  if (status === ScheduledEventStatus.completed) return 'Completed';
  if (status === ScheduledEventStatus.canceled) return 'Canceled';
  return 'Scheduled';
}

export function scheduledEventRecurrencePreset(
  rule: ScheduledEventRecurrenceRule | null | undefined
): ScheduledEventRecurrencePresetValue {
  if (!rule) return ScheduledEventRecurrencePreset.none;
  if (rule.frequency === ScheduledEventRecurrenceFrequency.daily)
    return ScheduledEventRecurrencePreset.daily;
  if (rule.frequency === ScheduledEventRecurrenceFrequency.weekly)
    return rule.interval === 2
      ? ScheduledEventRecurrencePreset.biweekly
      : ScheduledEventRecurrencePreset.weekly;
  if (rule.frequency === ScheduledEventRecurrenceFrequency.monthly)
    return ScheduledEventRecurrencePreset.monthly;
  if (rule.frequency === ScheduledEventRecurrenceFrequency.yearly)
    return ScheduledEventRecurrencePreset.yearly;
  return ScheduledEventRecurrencePreset.none;
}

export function scheduledEventRecurrenceLabel(
  rule: ScheduledEventRecurrenceRule | null | undefined
): string | null {
  const preset = scheduledEventRecurrencePreset(rule);
  if (preset === ScheduledEventRecurrencePreset.none) return null;
  if (preset === ScheduledEventRecurrencePreset.daily) return 'Repeats daily';
  if (preset === ScheduledEventRecurrencePreset.weekly) return 'Repeats weekly';
  if (preset === ScheduledEventRecurrencePreset.biweekly) return 'Repeats every 2 weeks';
  if (preset === ScheduledEventRecurrencePreset.monthly) return 'Repeats monthly';
  return 'Repeats yearly';
}

function isoTimestamp(value: string, field: string): string {
  const date = new Date(value);
  if (!value || !Number.isFinite(date.valueOf())) throw new Error(`Choose a valid ${field}.`);
  return date.toISOString();
}

function recurrenceRule(
  preset: ScheduledEventRecurrencePresetValue | undefined,
  scheduledStartTime: string
): ScheduledEventRecurrenceRule | null {
  if (!preset || preset === ScheduledEventRecurrencePreset.none) return null;
  const start = new Date(scheduledStartTime);
  const common = { start: start.toISOString(), end: null, interval: 1 as const };
  if (preset === ScheduledEventRecurrencePreset.daily)
    return { ...common, frequency: ScheduledEventRecurrenceFrequency.daily };
  if (preset === ScheduledEventRecurrencePreset.weekly) {
    return {
      ...common,
      frequency: ScheduledEventRecurrenceFrequency.weekly,
      by_weekday: [(start.getUTCDay() + 6) % 7]
    };
  }
  if (preset === ScheduledEventRecurrencePreset.biweekly) {
    return {
      ...common,
      frequency: ScheduledEventRecurrenceFrequency.weekly,
      interval: 2,
      by_weekday: [(start.getUTCDay() + 6) % 7]
    };
  }
  if (preset === ScheduledEventRecurrencePreset.monthly)
    return { ...common, frequency: ScheduledEventRecurrenceFrequency.monthly };
  return {
    ...common,
    frequency: ScheduledEventRecurrenceFrequency.yearly,
    by_month: [start.getUTCMonth() + 1],
    by_month_day: [start.getUTCDate()]
  };
}

export function scheduledEventPayload(draft: ScheduledEventDraft): Record<string, unknown> {
  const name = draft.name.trim();
  const description = draft.description.trim();
  if (!name) throw new Error('Give the scheduled event a name.');
  const scheduledStartTime = isoTimestamp(draft.startTime, 'start date and time');
  const scheduledEndTime = draft.endTime ? isoTimestamp(draft.endTime, 'end date and time') : null;
  if (scheduledEndTime && Date.parse(scheduledEndTime) <= Date.parse(scheduledStartTime)) {
    throw new Error('The end time must be later than the start time.');
  }

  if (draft.entityType !== ScheduledEventEntityType.external) {
    if (!draft.channelRef) throw new Error('Choose the channel where this event will happen.');
    return {
      channel_id: draft.channelRef,
      entity_metadata: null,
      name,
      privacy_level: 2,
      scheduled_start_time: scheduledStartTime,
      scheduled_end_time: scheduledEndTime,
      description: description || null,
      entity_type: draft.entityType,
      recurrence_rule: recurrenceRule(draft.recurrence, scheduledStartTime)
    };
  }

  const location = draft.location.trim();
  if (!location) throw new Error('Add the location or link for this external event.');
  if (!scheduledEndTime) throw new Error('Choose an end time for this external event.');
  return {
    channel_id: null,
    entity_metadata: { location },
    name,
    privacy_level: 2,
    scheduled_start_time: scheduledStartTime,
    scheduled_end_time: scheduledEndTime,
    description: description || null,
    entity_type: ScheduledEventEntityType.external,
    recurrence_rule: recurrenceRule(draft.recurrence, scheduledStartTime)
  };
}

/** Build a PATCH that never resends an unchanged, already-past start time. */
export function scheduledEventPatch(
  event: ScheduledEvent,
  draft: ScheduledEventDraft
): Record<string, unknown> {
  const next = scheduledEventPayload(draft);
  const previous: Record<string, unknown> = {
    channel_id: eventChannelRef(event),
    entity_metadata: event.entity_metadata,
    name: event.name,
    privacy_level: event.privacy_level,
    scheduled_start_time: new Date(event.scheduled_start_time).toISOString(),
    scheduled_end_time: event.scheduled_end_time
      ? new Date(event.scheduled_end_time).toISOString()
      : null,
    description: event.description,
    entity_type: event.entity_type,
    recurrence_rule: recurrenceRule(
      scheduledEventRecurrencePreset(event.recurrence_rule),
      event.scheduled_start_time
    )
  };
  return Object.fromEntries(
    Object.entries(next).filter(
      ([key, value]) => JSON.stringify(value) !== JSON.stringify(previous[key])
    )
  );
}

function collectionPath(guildRef: string): string {
  return `/guilds/${encodeURIComponent(guildRef)}/scheduled-events`;
}

function eventPath(guildRef: string, eventRef: string): string {
  return `${collectionPath(guildRef)}/${encodeURIComponent(eventRef)}`;
}

function eventImagePath(guildRef: string, event: ScheduledEvent): string {
  return `${eventPath(guildRef, scheduledEventRef(event))}/image`;
}

export async function uploadScheduledEventImage(
  guildRef: string,
  event: ScheduledEvent,
  file: File,
  onProgress: (progress: number) => void = () => undefined
): Promise<ScheduledEvent> {
  if (!file.type.toLowerCase().startsWith('image/')) throw new Error('Choose an image file.');
  if (file.size < 1 || file.size > 10 * 1024 * 1024) {
    throw new Error('Scheduled event cover images must be between 1 byte and 10 MiB.');
  }
  const path = eventImagePath(guildRef, event);
  const ticket = await api<UploadTicket>(`${path}/tickets`, {
    method: 'POST',
    body: JSON.stringify({
      filename: file.name || 'event-cover',
      content_type: file.type,
      size: file.size
    })
  });
  await uploadObject(ticket, file, onProgress);
  return completeScannedMediaResource(
    () =>
      api<ScheduledEvent | { status: string; attachment: { scan_status: string } }>(path, {
        method: 'PUT',
        body: JSON.stringify({ attachment_id: ticket.id })
      }),
    (value): value is ScheduledEvent => 'guild_id' in value && 'image' in value,
    {
      maxAttempts: 46,
      rejectedMessage: 'The event cover did not pass media processing. Choose another image.',
      timeoutMessage: 'The event cover is still processing. Try again shortly.'
    }
  );
}

export function deleteScheduledEventImage(
  guildRef: string,
  event: ScheduledEvent
): Promise<ScheduledEvent> {
  return api<ScheduledEvent>(eventImagePath(guildRef, event), { method: 'DELETE' });
}

export function listScheduledEvents(
  guildRef: string,
  signal?: AbortSignal
): Promise<ScheduledEvent[]> {
  return api<ScheduledEvent[]>(`${collectionPath(guildRef)}?with_user_count=true`, { signal });
}

export function createScheduledEvent(
  guildRef: string,
  draft: ScheduledEventDraft
): Promise<ScheduledEvent> {
  return api<ScheduledEvent>(collectionPath(guildRef), {
    method: 'POST',
    body: JSON.stringify(scheduledEventPayload(draft))
  });
}

export function editScheduledEvent(
  guildRef: string,
  event: ScheduledEvent,
  draft: ScheduledEventDraft
): Promise<ScheduledEvent> {
  const patch = scheduledEventPatch(event, draft);
  if (!Object.keys(patch).length) return Promise.resolve(event);
  return api<ScheduledEvent>(eventPath(guildRef, scheduledEventRef(event)), {
    method: 'PATCH',
    body: JSON.stringify(patch)
  });
}

export function transitionScheduledEvent(
  guildRef: string,
  event: ScheduledEvent,
  status: 2 | 3 | 4
): Promise<ScheduledEvent> {
  return api<ScheduledEvent>(eventPath(guildRef, scheduledEventRef(event)), {
    method: 'PATCH',
    body: JSON.stringify({ status })
  });
}

export function deleteScheduledEvent(guildRef: string, event: ScheduledEvent): Promise<void> {
  return api<void>(eventPath(guildRef, scheduledEventRef(event)), { method: 'DELETE' });
}

export function listScheduledEventUsers(
  guildRef: string,
  event: ScheduledEvent,
  options: { after?: string; limit?: number; signal?: AbortSignal } = {}
): Promise<ScheduledEventUser[]> {
  const query = new URLSearchParams({
    limit: String(options.limit ?? 100),
    with_member: 'true'
  });
  if (options.after) query.set('after', options.after);
  return api<ScheduledEventUser[]>(
    `${eventPath(guildRef, scheduledEventRef(event))}/users?${query.toString()}`,
    { signal: options.signal }
  );
}

export function setScheduledEventSubscription(
  guildRef: string,
  event: ScheduledEvent,
  subscribed: boolean
): Promise<void> {
  return api<void>(`${eventPath(guildRef, scheduledEventRef(event))}/users/@me`, {
    method: subscribed ? 'PUT' : 'DELETE'
  });
}

File: mobile/lib/src/domain/scheduled_events.dart
import 'package:kaede_mobile/src/core/errors.dart';
import 'package:kaede_mobile/src/core/refs.dart';
import 'package:kaede_mobile/src/domain/models.dart';

typedef ScheduledEventJson = Map<String, Object?>;

enum ScheduledEventStatus {
  scheduled(1),
  active(2),
  completed(3),
  canceled(4);

  const ScheduledEventStatus(this.wire);
  final int wire;

  static ScheduledEventStatus fromWire(Object? value) {
    final wire = value is num ? value.toInt() : int.tryParse('$value');
    return values.firstWhere(
      (candidate) => candidate.wire == wire,
      orElse: () => scheduled,
    );
  }
}

enum ScheduledEventEntityType {
  stage(1),
  voice(2),
  external(3);

  const ScheduledEventEntityType(this.wire);
  final int wire;

  static ScheduledEventEntityType fromWire(Object? value) {
    final wire = value is num ? value.toInt() : int.tryParse('$value');
    return values.firstWhere(
      (candidate) => candidate.wire == wire,
      orElse: () => external,
    );
  }
}

enum ScheduledEventRecurrencePreset {
  none,
  daily,
  weekly,
  biweekly,
  monthly,
  yearly,
}

enum ScheduledEventRecurrenceFrequency {
  yearly(0),
  monthly(1),
  weekly(2),
  daily(3);

  const ScheduledEventRecurrenceFrequency(this.wire);
  final int wire;
}

ScheduledEventRecurrencePreset scheduledEventRecurrencePreset(
  ScheduledEventJson? rule,
) {
  if (rule == null) return ScheduledEventRecurrencePreset.none;
  final frequency = (rule['frequency'] as num?)?.toInt();
  if (frequency == ScheduledEventRecurrenceFrequency.daily.wire) {
    return ScheduledEventRecurrencePreset.daily;
  }
  if (frequency == ScheduledEventRecurrenceFrequency.weekly.wire) {
    return (rule['interval'] as num?)?.toInt() == 2
        ? ScheduledEventRecurrencePreset.biweekly
        : ScheduledEventRecurrencePreset.weekly;
  }
  if (frequency == ScheduledEventRecurrenceFrequency.monthly.wire) {
    return ScheduledEventRecurrencePreset.monthly;
  }
  if (frequency == ScheduledEventRecurrenceFrequency.yearly.wire) {
    return ScheduledEventRecurrencePreset.yearly;
  }
  return ScheduledEventRecurrencePreset.none;
}

String? scheduledEventRecurrenceLabel(ScheduledEventJson? rule) =>
    switch (scheduledEventRecurrencePreset(rule)) {
      ScheduledEventRecurrencePreset.none => null,
      ScheduledEventRecurrencePreset.daily => 'Repeats daily',
      ScheduledEventRecurrencePreset.weekly => 'Repeats weekly',
      ScheduledEventRecurrencePreset.biweekly => 'Repeats every 2 weeks',
      ScheduledEventRecurrencePreset.monthly => 'Repeats monthly',
      ScheduledEventRecurrencePreset.yearly => 'Repeats yearly',
    };

final class GuildScheduledEvent {
  const GuildScheduledEvent({
    required this.ref,
    required this.guildRef,
    required this.creatorRef,
    required this.name,
    required this.startTime,
    required this.status,
    required this.entityType,
    required this.createdAt,
    required this.updatedAt,
    required this.version,
    this.channelRef,
    this.creator,
    this.description,
    this.endTime,
    this.location,
    this.imageHash,
    this.recurrenceRule,
    this.userCount = 0,
    this.meSubscribed = false,
  });

  factory GuildScheduledEvent.fromJson(ScheduledEventJson json) {
    final domain = Domain('${json['origin_domain']}');
    final guildDomain = Domain('${json['guild_domain']}');
    final channelId = json['channel_id'];
    final channelDomain = json['channel_domain'];
    final creatorJson = json['creator'];
    final metadata = json['entity_metadata'];
    return GuildScheduledEvent(
      ref: EntityRef(Snowflake('${json['id']}'), domain),
      guildRef: EntityRef(Snowflake('${json['guild_id']}'), guildDomain),
      channelRef: channelId == null || channelDomain == null
          ? null
          : EntityRef(Snowflake('$channelId'), Domain('$channelDomain')),
      creatorRef: EntityRef(
        Snowflake('${json['creator_id']}'),
        Domain('${json['creator_domain']}'),
      ),
      creator: creatorJson is Map
          ? KaedeUser.fromJson(Map<String, Object?>.from(creatorJson))
          : null,
      name: '${json['name']}',
      description: json['description'] as String?,
      startTime: DateTime.parse('${json['scheduled_start_time']}').toLocal(),
      endTime: json['scheduled_end_time'] == null
          ? null
          : DateTime.parse('${json['scheduled_end_time']}').toLocal(),
      status: ScheduledEventStatus.fromWire(json['status']),
      entityType: ScheduledEventEntityType.fromWire(json['entity_type']),
      location: metadata is Map ? metadata['location'] as String? : null,
      imageHash: json['image'] as String?,
      recurrenceRule: json['recurrence_rule'] is Map
          ? Map<String, Object?>.from(json['recurrence_rule']! as Map)
          : null,
      createdAt: DateTime.parse('${json['created_at']}').toUtc(),
      updatedAt: DateTime.parse('${json['updated_at']}').toUtc(),
      version: '${json['version']}',
      userCount: (json['user_count'] as num?)?.toInt() ?? 0,
      meSubscribed: json['me_subscribed'] == true,
    );
  }

  final EntityRef ref;
  final EntityRef guildRef;
  final EntityRef? channelRef;
  final EntityRef creatorRef;
  final KaedeUser? creator;
  final String name;
  final String? description;
  final DateTime startTime;
  final DateTime? endTime;
  final ScheduledEventStatus status;
  final ScheduledEventEntityType entityType;
  final String? location;
  final String? imageHash;
  final ScheduledEventJson? recurrenceRule;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String version;
  final int userCount;
  final bool meSubscribed;

  GuildScheduledEvent copyWith({int? userCount, bool? meSubscribed}) =>
      GuildScheduledEvent(
        ref: ref,
        guildRef: guildRef,
        channelRef: channelRef,
        creatorRef: creatorRef,
        creator: creator,
        name: name,
        description: description,
        startTime: startTime,
        endTime: endTime,
        status: status,
        entityType: entityType,
        location: location,
        imageHash: imageHash,
        recurrenceRule: recurrenceRule,
        createdAt: createdAt,
        updatedAt: updatedAt,
        version: version,
        userCount: userCount ?? this.userCount,
        meSubscribed: meSubscribed ?? this.meSubscribed,
      );
}

final class ScheduledEventSubscriber {
  const ScheduledEventSubscriber({
    required this.eventRef,
    required this.user,
    required this.subscribedAt,
    this.member,
  });

  factory ScheduledEventSubscriber.fromJson(ScheduledEventJson json) {
    final user = KaedeUser.fromJson(
      Map<String, Object?>.from(json['user']! as Map),
    );
    final rawMember = json['member'];
    return ScheduledEventSubscriber(
      eventRef: EntityRef(
        Snowflake('${json['guild_scheduled_event_id']}'),
        Domain('${json['guild_scheduled_event_domain']}'),
      ),
      user: user,
      member: rawMember is Map
          ? GuildMember.fromJson(Map<String, Object?>.from(rawMember))
          : null,
      subscribedAt: DateTime.parse('${json['subscribed_at']}').toUtc(),
    );
  }

  final EntityRef eventRef;
  final KaedeUser user;
  final GuildMember? member;
  final DateTime subscribedAt;
}

final class ScheduledEventDraft {
  const ScheduledEventDraft({
    required this.name,
    required this.description,
    required this.entityType,
    required this.startTime,
    this.channelRef,
    this.location = '',
    this.endTime,
    this.recurrence = ScheduledEventRecurrencePreset.none,
  });

  factory ScheduledEventDraft.fromEvent(GuildScheduledEvent event) =>
      ScheduledEventDraft(
        name: event.name,
        description: event.description ?? '',
        entityType: event.entityType,
        channelRef: event.channelRef,
        location: event.location ?? '',
        startTime: event.startTime,
        endTime: event.endTime,
        recurrence: scheduledEventRecurrencePreset(event.recurrenceRule),
      );

  final String name;
  final String description;
  final ScheduledEventEntityType entityType;
  final EntityRef? channelRef;
  final String location;
  final DateTime startTime;
  final DateTime? endTime;
  final ScheduledEventRecurrencePreset recurrence;

  ScheduledEventJson? _recurrenceJson(DateTime start) {
    if (recurrence == ScheduledEventRecurrencePreset.none) return null;
    final utc = start.toUtc();
    final common = <String, Object?>{
      'start': utc.toIso8601String(),
      'end': null,
      'interval': 1,
    };
    return switch (recurrence) {
      ScheduledEventRecurrencePreset.none => null,
      ScheduledEventRecurrencePreset.daily => <String, Object?>{
          ...common,
          'frequency': ScheduledEventRecurrenceFrequency.daily.wire,
        },
      ScheduledEventRecurrencePreset.weekly => <String, Object?>{
          ...common,
          'frequency': ScheduledEventRecurrenceFrequency.weekly.wire,
          'by_weekday': <int>[utc.weekday - 1],
        },
      ScheduledEventRecurrencePreset.biweekly => <String, Object?>{
          ...common,
          'frequency': ScheduledEventRecurrenceFrequency.weekly.wire,
          'interval': 2,
          'by_weekday': <int>[utc.weekday - 1],
        },
      ScheduledEventRecurrencePreset.monthly => <String, Object?>{
          ...common,
          'frequency': ScheduledEventRecurrenceFrequency.monthly.wire,
        },
      ScheduledEventRecurrencePreset.yearly => <String, Object?>{
          ...common,
          'frequency': ScheduledEventRecurrenceFrequency.yearly.wire,
          'by_month': <int>[utc.month],
          'by_month_day': <int>[utc.day],
        },
    };
  }

  ScheduledEventJson toCreateJson() {
    final cleanName = name.trim();
    final cleanDescription = description.trim();
    final cleanLocation = location.trim();
    if (cleanName.isEmpty) {
      throw const UserInputException('Give the scheduled event a name.');
    }
    if (endTime != null && !endTime!.isAfter(startTime)) {
      throw const UserInputException(
        'The end time must be later than the start time.',
      );
    }
    if (entityType != ScheduledEventEntityType.external && channelRef == null) {
      throw const UserInputException(
        'Choose the channel where this event will happen.',
      );
    }
    if (entityType == ScheduledEventEntityType.external) {
      if (cleanLocation.isEmpty) {
        throw const UserInputException(
          'Add the location or link for this external event.',
        );
      }
      if (endTime == null) {
        throw const UserInputException(
          'Choose an end time for this external event.',
        );
      }
    }
    return <String, Object?>{
      'channel_id': entityType != ScheduledEventEntityType.external
          ? channelRef!.wire
          : null,
      'entity_metadata': entityType == ScheduledEventEntityType.external
          ? <String, Object?>{'location': cleanLocation}
          : null,
      'name': cleanName,
      'privacy_level': 2,
      'scheduled_start_time': startTime.toUtc().toIso8601String(),
      'scheduled_end_time': endTime?.toUtc().toIso8601String(),
      'description': cleanDescription.isEmpty ? null : cleanDescription,
      'entity_type': entityType.wire,
      'recurrence_rule': _recurrenceJson(startTime),
    };
  }

  ScheduledEventJson patchFor(GuildScheduledEvent event) {
    final next = toCreateJson();
    final previous = <String, Object?>{
      'channel_id': event.channelRef?.wire,
      'entity_metadata': event.entityType == ScheduledEventEntityType.external
          ? <String, Object?>{'location': event.location}
          : null,
      'name': event.name,
      'privacy_level': 2,
      'scheduled_start_time': event.startTime.toUtc().toIso8601String(),
      'scheduled_end_time': event.endTime?.toUtc().toIso8601String(),
      'description': event.description,
      'entity_type': event.entityType.wire,
      'recurrence_rule':
          ScheduledEventDraft.fromEvent(event)._recurrenceJson(event.startTime),
    };
    next.removeWhere((key, value) => _sameJsonValue(value, previous[key]));
    return next;
  }
}

bool _sameJsonValue(Object? left, Object? right) {
  if (left is Map && right is Map) {
    if (left.length != right.length) return false;
    return left.entries.every(
      (entry) =>
          right.containsKey(entry.key) &&
          _sameJsonValue(entry.value, right[entry.key]),
    );
  }
  if (left is List && right is List) {
    if (left.length != right.length) return false;
    for (var index = 0; index < left.length; index += 1) {
      if (!_sameJsonValue(left[index], right[index])) return false;
    }
    return true;
  }
  return left == right;
}
